L1Trigger/L1THGCal/src/be_algorithms/HGCalShowerShape.cc

```diff
@@ -0,0 +1,554 @@
+        if(clu->subdetId()==HGCHEF) layer=clu->layer()+nLayersEE;//FH
+
+        if(layer<firstLayer) firstLayer=layer;
+
+    }
+
+    return firstLayer;
+
+}
+
+
+
+
+const float HGCalShowerShape::sigmaEtaEtaTot(const l1t::HGCalMulticluster& c3d){
+
+    const edm::PtrVector<l1t::HGCalCluster> clustersPtrs = c3d.constituents();
+
+    std::vector<float> tc_energy ; 
+    std::vector<float> tc_eta ;
+
+    for(const auto& clu : clustersPtrs){
+
+        const edm::PtrVector<l1t::HGCalTriggerCell> triggerCells = clu->constituents();
+        unsigned int ncells = triggerCells.size();
+
+        for(unsigned int itc=0; itc<ncells;itc++){
+
+            l1t::HGCalTriggerCell thistc = *triggerCells[itc];
+
+            tc_energy.emplace_back(thistc.energy());
+            tc_eta.emplace_back(thistc.eta());
+                
+        }
+
+    }
+
+    float SeeTot = sigmaEtaEta(tc_energy,tc_eta);
+
+    tc_energy.clear();
+    tc_eta.clear();
+
+    return SeeTot;
+
+}
+
+
+
+
+const float HGCalShowerShape::sigmaPhiPhiTot(const l1t::HGCalMulticluster& c3d){
+
+
+    const edm::PtrVector<l1t::HGCalCluster> clustersPtrs = c3d.constituents();
+
+    std::vector<float> tc_energy ; 
+    std::vector<float> tc_phi ;
+
+    for(const auto& clu : clustersPtrs){
+
+        const edm::PtrVector<l1t::HGCalTriggerCell> triggerCells = clu->constituents();
+        unsigned int ncells = triggerCells.size();
+
+        for(unsigned int itc=0; itc<ncells;itc++){
+
+            l1t::HGCalTriggerCell thistc = *triggerCells[itc];
+
+            tc_energy.emplace_back(thistc.energy());
+            tc_phi.emplace_back(thistc.phi());
+                
+        }
+    }
+
+    float SppTot = sigmaPhiPhi(tc_energy,tc_phi);
+
+    tc_energy.clear();
+    tc_phi.clear();
+
+    return SppTot;
+
+}
+
+
+
+const float HGCalShowerShape::sigmaEtaEtaMax(const l1t::HGCalMulticluster& c3d){
+
+    const edm::PtrVector<l1t::HGCalCluster> clustersPtrs = c3d.constituents();
+
+    std::vector<int> layer ; // Size : ncl2D
+    std::vector<int> subdetID ;
+    std::vector<int> nTC ;
+    std::vector<float> tc_energy ; 
+    std::vector<float> tc_eta ;
+
+    for(const auto& clu : clustersPtrs){
+
+        layer.emplace_back(clu->layer());
+        subdetID.emplace_back(clu->subdetId());
+
+        const edm::PtrVector<l1t::HGCalTriggerCell> triggerCells = clu->constituents();
+        unsigned int ncells = triggerCells.size();
+        nTC.emplace_back(ncells);
+
+        for(unsigned int itc=0; itc<ncells;itc++){
+
+            l1t::HGCalTriggerCell thistc = *triggerCells[itc];
+
+            tc_energy.emplace_back(thistc.energy());
+            tc_eta.emplace_back(thistc.eta());
+                
+        }
+
+    }
+
+    int ncl2D=layer.size();
+
+    float SigmaEtaEtaMax=0;
+
+    std::vector<float> energy_layer ;
+    std::vector<float> eta_layer ;
+
+    int nLayersMax=40;//EE+FH
+    int nLayersEE=28;
+
+    for(int ilayer=0;ilayer<nLayersMax;ilayer++){   //Loop on HGCal layers
```

Comment:
If the energies and etas were already grouped by layer (as mentioned above), the code below would be much much simpler.